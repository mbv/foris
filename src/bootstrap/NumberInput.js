/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";

import PropTypes from "prop-types";
import { Input } from "./Input";

NumberInput.propTypes = {
    /** Field label. */
    label: PropTypes.string.isRequired,
    /** Error message. */
    error: PropTypes.string,
    /** Help text message. */
    helpText: PropTypes.string,
    /** Number value. */
    value: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number,
    ]),
    /** Function called when value changes. */
    onChange: PropTypes.func.isRequired,
    /** Additional description dispaled to the right of input value. */
    inlineText: PropTypes.string,
};

NumberInput.defaultProps = {
    value: 0,
};

export function NumberInput({ onChange, inlineText, ...props }) {
    return (
        <Input type="number" onChange={onChange} {...props}>
            <div className="input-group-append">
                {inlineText && <p className="input-group-text">{inlineText}</p>}
                <button
                    type="button"
                    className="btn btn-outline-secondary"
                    onClick={() => onChange({ target: { value: props.value + 1 } })}
                    aria-label="Increase"
                >
                    <i className="fas fa-plus" />
                </button>
                <button
                    type="button"
                    className="btn btn-outline-secondary"
                    onClick={() => onChange({ target: { value: props.value - 1 } })}
                    aria-label="Decrease"
                >
                    <i className="fas fa-minus" />
                </button>
            </div>
        </Input>
    );
}
