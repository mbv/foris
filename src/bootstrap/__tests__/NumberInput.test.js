/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";

import { render, fireEvent } from "customTestRender";

import { NumberInput } from "../NumberInput";

describe("<NumberInput/>", () => {
    const onChangeMock = jest.fn();
    let container;
    let getByLabelText;

    beforeEach(() => {
        ({ container, getByLabelText } = render(
            <NumberInput
                label="Test label"
                helpText="Some help text"
                value={1}
                onChange={onChangeMock}
            />
        ));
    });

    it("Render number input", () => {
        expect(container.firstChild).toMatchSnapshot();
    });

    it("Increase number with button", () => {
        const increaseButton = getByLabelText("Increase");
        fireEvent.click(increaseButton);
        expect(onChangeMock).toHaveBeenCalledWith({"target": {"value": 2}});
    });

    it("Decrease number with button", () => {
        const decreaseButton = getByLabelText("Decrease");
        fireEvent.click(decreaseButton);
        expect(onChangeMock).toHaveBeenCalledWith({"target": {"value": 0}});
    });
});
