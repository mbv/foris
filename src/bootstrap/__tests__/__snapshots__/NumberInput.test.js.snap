// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NumberInput/> Render number input 1`] = `
<div
  class="col-sm-12 offset-lg-1 col-lg-10"
>
  <div
    class="form-group"
  >
    <label
      for="1"
    >
      Test label
    </label>
    <div
      class="input-group"
    >
      <input
        class="form-control"
        id="1"
        type="number"
        value="1"
      />
      <div
        class="input-group-append"
      >
        <button
          aria-label="Increase"
          class="btn btn-outline-secondary"
          type="button"
        >
          <i
            class="fas fa-plus"
          />
        </button>
        <button
          aria-label="Decrease"
          class="btn btn-outline-secondary"
          type="button"
        >
          <i
            class="fas fa-minus"
          />
        </button>
      </div>
    </div>
    <small
      class="form-text text-muted"
    >
      Some help text
    </small>
  </div>
</div>
`;
